### PR TITLE
[algo] fix: loss metrics inflated by num_mini_batches * ppo_epochs

### PR DESCRIPTION
**Pull request by dubin555**
### What does this PR do?

Fixes a bug where `actor/pg_loss`, `actor/kl_loss`, and `critic/vf_loss` metrics are inflated by a factor of `num_mini_batches × ppo_epochs`.

**Root cause**: PR #4711 changed loss metric accumulation from list-based (`append_to_dict`) to float-scalar (`+= ... * loss_scale_factor`) to fix double-scaling. However, this introduced a new issue: the float scalar accumulates across ALL mini-batches and epochs without resetting. When `reduce_metrics` later calls `np.mean()` on the value, a scalar passes through as-is (`np.mean(float) == float`), so the final metric is a **sum** instead of a **mean**.

**Impact**: Logged metrics only — training dynamics are unaffected. With `num_mini_batches=4, ppo_epochs=2`, the reported loss is 8× the actual per-mini-batch value, making it difficult to compare runs or detect real training issues.

**Fix**: Accumulate loss within each mini-batch using local variables (`mini_batch_pg_loss`, `mini_batch_kl_loss`, `mini_batch_vf_loss`), then `append_to_dict` at the mini-batch boundary. This way `reduce_metrics` correctly averages across all mini-batch iterations.

### Checklist Before Starting

- [x] Search for similar PRs: https://github.com/verl-project/verl/pulls?q=is%3Apr+loss+metric+inflation
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

6 unit tests in `tests/metric/test_loss_metric_inflation.py`:

- `test_single_mini_batch_single_epoch` — old and new agree when there's no multi-iteration
- `test_two_mini_batches_old_inflated` — old=5.0 (2× inflated), new=2.5 (correct)
- `test_two_epochs_old_inflated` — old=3.0 (2× inflated), new=1.5 (correct)
- `test_two_mini_batches_two_epochs` — old=4.0 (4× inflated), new=1.0 (correct)
- `test_inflation_factor` — verifies `old/new == num_mini_batches * ppo_epochs`
- `test_metrics_dict_types` — all metric values are lists after fix (no mixed float/list)

### Design & Code Changes

**`verl/workers/actor/dp_actor.py`**:
- Initialize `metrics = {}` (was `{"actor/pg_loss": 0.0, "actor/kl_loss": 0.0}`)
- Add `mini_batch_pg_loss` and `mini_batch_kl_loss` local variables, scoped per mini-batch iteration
- Accumulate into locals during micro-batch loop, then `append_to_dict` at mini-batch boundary

**`verl/workers/critic/dp_critic.py`**:
- Same pattern: `mini_batch_vf_loss` local variable, scoped per mini-batch iteration

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting)
- [x] Add unit tests to cover the fix
- [ ] CI request in Slack

### CODE REVIEW

File: verl/trainer/ppo/ray_trainer.py [hunk 1]
```diff
@@ -1061,6 +1066,19 @@ def _load_checkpoint(self):
         else:
             print(f"Warning: No dataloader state found at {dataloader_local_path}, will start from scratch")
 
+        # load KL controller state
+        if self.config.algorithm.use_kl_in_reward:
+            kl_ctrl_path = os.path.join(global_step_folder, "kl_ctrl.pt")
+            if os.path.exists(kl_ctrl_path):
+                kl_state = torch.load(kl_ctrl_path, weights_only=False)
+                self.kl_ctrl_in_reward.value = kl_state["value"]
```
[the remark this answers is not in the document]
Comment (dubin555) [reply verl/trainer/ppo/ray_trainer.py]: Good catch\! Updated to use `weights_only=True` by wrapping the value in `torch.tensor()` on save and calling `.item()` on load. Since we're only persisting a single float, there's no need for pickle.

File: tests/test_adaptive_kl_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            kl_ctrl_path = os.path.join(tmpdir, "kl_ctrl.pt")
+            torch.save({"value": ctrl.value}, kl_ctrl_path)
+
+            # Simulate resume: create a fresh controller (as __init__ would)
+            new_ctrl = AdaptiveKLController(init_kl_coef=init_coef, target_kl=1.0, horizon=10000)
+            assert new_ctrl.value == init_coef, "Fresh controller should have init value"
+
+            # Load (matching the fix in ray_trainer.py _load_checkpoint)
+            kl_state = torch.load(kl_ctrl_path, weights_only=False)
+            new_ctrl.value = kl_state["value"]
```
[the remark this answers is not in the document]
Comment (dubin555) [reply tests/test_adaptive_kl_checkpoint.py]: Updated the test as well to match the safer implementation. Thanks!

File: verl/trainer/ppo/ray_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dubin555) [reply verl/trainer/ppo/ray_trainer.py]: These comments were about the previous version of this PR (AdaptiveKLController checkpoint persistence). The branch has since been rebased — the current diff addresses loss metric inflation in `dp_actor.py` and `dp_critic.py`, and no longer contains any `torch.load` calls. I've updated the PR title and description to reflect the current changes. Apologies for the confusion\!

**Comment by dubin555** [comment]
Closing this PR — the upstream codebase has since undergone a major refactoring ([#6067](https://github.com/verl-project/verl/pull/6067)) that deprecated and removed the legacy `dp_actor.py` and `dp_critic.py` workers entirely. The loss computation logic has been migrated to `verl/workers/utils/losses.py` with a new `Metric` class using explicit `AggregationType` (SUM/MEAN), which architecturally eliminates the float `+=` accumulation pattern that caused the metric inflation.

Since the target files no longer exist in the current codebase, this fix is no longer applicable. Thanks for the review feedback on the earlier iteration!
